File: Self-practice/evenodd.py
MODULO = 10 ** 9 + 7

def f(x, dp):
    if x not in dp:
        odd, even = (x + 1) // 2, x // 2
        dp[x] = (even + 2 * (odd - 1) + f(odd, dp) + f(even, dp)) % MODULO
    return dp[x]

l, r = map(int, input().split())
dp = {0: 0, 1: 0}
print((f(r, dp) - f(l - 1, dp) + MODULO) % MODULO)

File: Self-practice/evenodd.cpp
#include<bits/stdc++.h>
using namespace std;

constexpr int MODULO = 1e9 + 7;

long long f(long long x, unordered_map<long long, long long> &dp) {
    if (!dp.count(x)) {
        auto even = x / 2, odd = (x + 1) / 2;
        dp[x] = (even + 2 * (odd - 1) + f(odd, dp) + f(even, dp)) % MODULO;
    }

    return dp[x];
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    long long l, r;
    cin >> l >> r;

    unordered_map<long long, long long> dp{{0, 0}, {1, 0}};
    cout << (f(r, dp) - f(l - 1, dp) + MODULO) % MODULO;
}
